draft-ietf-quic-datagram.md

```diff
@@ -278,13 +278,12 @@ dropped by the receiver if the receiver cannot process them.
 
 DATAGRAM frames employ the QUIC connection's congestion controller. As a
 result, a connection might be unable to send a DATAGRAM frame generated by the
-application until the congestion controller allows it
-{{!RFC9002}}. The sender implementation MUST either delay
-sending the frame until the controller allows it or drop the frame without
-sending it (at which point it MAY notify the application). Implementations
-that use packet pacing SHOULD support delaying the transmission of DATAGRAM
-frames for at least the time it takes to send the paced packets allowed by the
-congestion controller to avoid dropping frames excessively.
+application until the congestion controller allows it {{!RFC9002}}. The sender
+implementation MUST either delay sending the frame until the controller allows
+it or drop the frame without sending it (at which point it MAY notify the
+application). Implementations that use packet pacing ({{Section 7.7 of RFC9002}})
+SHOULD take care to not drop DATAGRAM frames due to the transmission delays
+used for pacing, to avoid excessive packet drops.
```

Comment:
Incorrect pacing would cause datagram frame loss, not packet loss. Here's a slightly tweaked phrasing:

```suggestion
application). Implementations that use packet pacing ({{Section 7.7 of RFC9002}})
SHOULD take care to avoid excessively dropping DATAGRAM frames due to the
transmission delays used for pacing.
```

That said, as an implementer I'm not sure what this sentence tells me: "take care" is very vague. What's a concrete solution? Keeping pacing credits available in case datagrams come in?

Reply:
Realistically, it's "don't drop a frame just because you're pacing". This came from https://github.com/quicwg/datagram/issues/16 originally.

@DavidSchinazi happy to take any suggested text here.

Reply:
If the application passes a datagram to the transport and the transport can't send it right away due to pacing, what's the transport supposed to do? If I shouldn't drop it what am I supposed to do instead?